### ISSUE TITLE
"native-assets" feature should surface stdout/stderr from build.dart

### ISSUE THREAD

**Issue by craiglabenz** [dart-lang/sdk#53732]
When using the new native assets feature (dart-lang/sdk#50565 and flutter/flutter#129757), the code that invokes `build.dart` seems to be burying its stdout and stderr.

I think it makes sense that end developers will not want to be bothered by log statements from these build steps, but seeing output / logging statements is important _when writing or debugging build.dart_.

I'm not sure how to differentiate between these two scenarios, but an escape hatch for `build.dart` authors would be a significant DX improvement.

cc @dcharkes @mit-mit

**Comment by dcharkes** [comment]
@mit-mit I believe I hid the logs on your request in the Dart SDK. 😄

**Comment by mit-mit** [comment]
My feedback was focused on the consumption use-case where it sounds like @craiglabenz agrees we should not show it.

Could we support seeing the output during development with a CLI flag or something like that?

**Comment by mosuem** [comment]
Even for end users who are not debugging their own build hooks, we should definitely show some kind of progress update IMO. If a user imports a package with a build or link hook invoking expensive operations, such as compiling a large native library from scratch, running `dart run` may take several minutes. For the user, it is invisible what is causing this. 

Some possible options off the top of my head:
#### 1: Show no progress (current situation)
- Pros: Clean empty output for `dart run`.
- Cons: Builds may take long time, involving multiple steps, the user is left in the dark about this. I personally always wonder if something crashed in my hook.

#### 2: Show `Running build hooks...`
- Pros: Minimal impact on output.
- Cons: It's not clear which package is causing the delay, or how long it will take.

#### 3: Show `Running build hook for ${package.name}...`
- Pros: User knows exactly who is causing the delay.
- Cons: Might be quite a lot of lines, if many packages have build hooks.

#### 4: Let the package show it's own message
- Pros: Packages which have fast-running build hooks could opt to not show a message, while longer-running ones can give detailed updates, up to progress bars for downloads etc.
- Cons: Uncontrollable amount of output in the console for the end user. Worst case: `package:ads_for_company` pays package authors to include it's build.dart in packages, which is just a delay and some advertising text :upside_down_face: (I hope this is against some policy of ours).

Opinions? @dcharkes @mit-mit @mariamhas @craiglabenz

**Comment by dcharkes** [comment]
I fear 3 and 4 might be too spammy.

Maybe we can do 2 in a similar style to `Building package executable... (2.1s)`. E.g. we keep updating the seconds so that it shows the process isn't hanging.

If we want to make it less spammy we can show it only after 5 seconds.

And I believe we should do it in conjunction with https://github.com/dart-lang/native/issues/97:

`Running build hooks. For logs see .dart_tool/native_assets_builder/logs/20240730164559.txt. (10.1s)`

**Comment by mit-mit** [comment]
I suggest 2) in the standard mode, and 3) when `dart` has been passed `--verbose`

**Comment by dcharkes** [comment]
> I suggest 2) in the standard mode, and 3) when `dart` has been passed `--verbose`

If we want to align with Flutter `-v`, then `--verbose` should stream all stdout and stderr from all hooks. So that would be option 4.
(But `flutter -v` is so extremely verbose that it feels useless. So maybe we shouldn't be aiming to that.)

**Comment by MichealReed** [comment]
4: Let the package show it's own message is a must in my opinion.

I opened https://github.com/dart-lang/native/issues/2075 before I discover this issue. I can close it if you only want to track this here. For libraries with long build times it's very important to see the full output of the build. Some build systems will not immediately terminate the build on errors. If a build system hangs there is no way to determine this without streaming output. I am building Google Dawn via cmake, these builds can take a significant amount of time.

--verbose flags should provide streaming output of the full build on both dart and flutter, otherwise it will be very difficult to troubleshoot deep compilations on varying toolchains.

**Comment by rainyl** [comment]
> If we want to align with Flutter `-v`, then `--verbose` should stream all stdout and stderr from all hooks. So that would be option 4. (But `flutter -v` is so extremely verbose that it feels useless. So maybe we shouldn't be aiming to that.)

For debugging purpose, `dart --verbose` should be able to print all stdout messages. As a plugin developer, `flutter -v` is useful when debugging some package building issues, like observing CMake build logs and make sure it's correct or fix it if failed.

**Comment by dcharkes** [comment]
I think we actually have two distinct issues we need to solve.

1. Not burying stderr/stdout of hooks of packages currently under development.
   "My program isn't working and I want ease of use to fix it. I can't find logs."
   Tracked in this issue.
2. Progress updates that packages are running in general.
    "Is my computer hanging?"
    Tracked in https://github.com/dart-lang/native/issues/2439.

The progress updates I believe are _only_ relevant when an interactive terminal is used. And progress updates are a separate stream of information compared to logs.

Assuming we're fixing progress updates as a separate issue, logs shouldn't be misused for progress updates.

The main question then becomes: when does a developer want do debug a certain hook and see its logs? Some heuristics:

* Definitely if the hook run fails/crashes - or if the stderr from a hook is not empty.
* Probably not for packages that are pulled from pub.dev, they are likely not owned by the developer.
  * Unless, the hook of that package has some user-defines, and the developer has trouble passing in the right user-defines.
* Possibly the hook of the root package, and possibly the hooks packages in the same pub workspace and the path dependencies. The developer is developing these.
  * However, the user might not be trying to make the hooks behave at the moment. So just streaming stderr and stdout of these always is too spammy.

So, I don't believe a heuristic will work reliably.

> Could we support seeing the output during development with a CLI flag or something like that?

```
$ dart --debug-hooks run
```

And potentially use an option instead of flag so that the value can be a package name or package names:

```
$ dart --debug-hooks=foo,bar run
```

This would then be a generic Dart option so that it would apply to any dartdev commands. (Currently, `run`, `test`, `build`, and `install`. And it'd be a no-op for the other commands.)

We'd want something similar in Flutter:

```
$ flutter --debug-hooks=foo,bar run -dmacos
```

However, due to `flutter run` not being able to see `flutter assemble`'s output due to those being burried in the logs of the xcodebuild/msvcbuild/Gradlebuild it might not be so easy to achieve that.

* https://github.com/flutter/flutter/issues/136386
* https://github.com/flutter/flutter/issues/147544

**Comment by MichealReed** [comment]
Great idea to control with a flag. Maybe the verbose flag should be extended for this case instead of introducing a new flag? If I add -v to flutter or dart I want all console output, even for packages. If you cannot pipe all output to verbose, maybe -vp or --verbose-packages=all,foo,bar?
